File: epi_judge_python/pascal_triangle.py
from typing import List

from test_framework import generic_test


def generate_pascal_triangle(n: int) -> List[List[int]]:
    if not n:
        return []
    res = [[1]]
    for i in range(n - 1):
        row = [1]
        for j in range(len(res[i]) - 1):
            row.append(res[i][j] + res[i][j + 1])
        row.append(1)
        res.append(row)
    return res


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "pascal_triangle.py", "pascal_triangle.tsv", generate_pascal_triangle
        )
    )

File: epi_judge_cpp/pascal_triangle.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;
vector<vector<int>> GeneratePascalTriangle(int num_rows) {
  if (num_rows == 0) {
    return {};
  }
  vector<vector<int>> pascal;
  pascal.push_back({1});
  for (auto i = 1; i < num_rows; i++) {
    auto &prev = pascal.back();
    vector<int> row;
    row.emplace_back(prev[0]);
    for (auto it = prev.begin(); it != prev.end() - 1; it++) {
      row.emplace_back(*it + *(it + 1));
    }
    row.emplace_back(prev[0]);
    pascal.push_back(row);
  }

  return pascal;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"num_rows"};
  return GenericTestMain(args, "pascal_triangle.cc", "pascal_triangle.tsv",
                         &GeneratePascalTriangle, DefaultComparator{},
                         param_names);
}
